### ISSUE TITLE
Google CloudSQL authorized networks not populated

### ISSUE THREAD

**Issue by coreone** [hashicorp/terraform#4222]
I have been using the new Google CloudSQL resource in Terraform, which all seems to work great, except for the `authorized_networks` configuration.  When I fill in this section with one or more networks, it blanks out all the authorized networks on the Google side, even though the **plan** and **apply** steps both seem to suggest that the correct changes are being made.  I started this discussion at:

https://groups.google.com/d/msgid/terraform-tool/88db0e72-8b10-493e-9fa9-fb8ec66c3682%40googlegroups.com

My configuration is as follows (identifying information removed):
```
resource "google_sql_database_instance" "mysql" {
    name = "mysql${format("%02d", count.index+1)}"
    count = "${var.num_cloudsql_instances}"
    region = "${var.cloudsql_region}"
    database_version = "${var.cloudsql_version}"

    settings {
        activation_policy = "${var.cloudsql_activation_policy}"
        pricing_plan = "${var.cloudsql_pricing_plan}"
        replication_type = "${var.cloudsql_replication_type}"
        tier = "D16"

        backup_configuration {
            binary_log_enabled = false
            enabled = true
            start_time = "06:00"
        }

        ip_configuration {
            ipv4_enabled = true
            # require_ssl = true
            authorized_networks = {
                name = "The Network"
                value = "x.x.x.x/x"
            }
            authorized_networks = {
                name = "${google_compute_instance.computename.0.name}"
                value = "${google_compute_address.computename.0.address}/32"
            }
        }
    }
}
```
When I run the plan, I get:
```
~ google_sql_database_instance.mysql
    settings.0.ip_configuration.0.authorized_networks.#:       "0" => "2"
    settings.0.ip_configuration.0.authorized_networks.0.name:  "" => "The Network"
    settings.0.ip_configuration.0.authorized_networks.0.value: "" => "x.x.x.x/x"
    settings.0.ip_configuration.0.authorized_networks.1.name:  "" => "computename"
    settings.0.ip_configuration.0.authorized_networks.1.value: "" => "x.x.x.x/32"
```

And then apply:
```
google_sql_database_instance.mysql: Modifying...
  settings.0.ip_configuration.0.authorized_networks.#:       "0" => "2"
  settings.0.ip_configuration.0.authorized_networks.0.name:  "" => "The Network"
  settings.0.ip_configuration.0.authorized_networks.0.value: "" => "x.x.x.x/x"
  settings.0.ip_configuration.0.authorized_networks.1.name:  "" => "computename"
  settings.0.ip_configuration.0.authorized_networks.1.value: "" => "x.x.x.x/32"
google_sql_database_instance.mysql: Modifications complete

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.
```

However, all the authorized networks are removed, and neither of these is added.

Let me know if you need any further information.

**Comment by coreone** [comment]
@jen20 This is becoming a real issue as certain unrelated setting updates to CloudSQL now wipe existing authorized networks, even if the *authorized_networks* setting is not being used in the *google_sql_database_instance* resource.

**Comment by kafis** [comment]
Any progress on this?

**Comment by phinze** [comment]
Hey everybody,

Sorry that this one has hung out for so long! I'm working on reproducing it now, and I'm having trouble coming up with a repro case on this side. 

Here is the config I'm working with currently ( based off @hjfbynara's example in #5630 )

```tf
resource "google_sql_database_instance" "instance" {
  name   = "tf-lw-0612"
  region = "us-central"

  settings {
    tier                   = "D0"
    crash_safe_replication = false
    replication_type       = "ASYNCHRONOUS"

    location_preference {
      zone = "us-central1-f"
    }

    ip_configuration {
      ipv4_enabled = "true"

      authorized_networks {
        value           = "108.12.12.0/24"
        name            = "misc"
        expiration_time = "2017-11-15T16:19:00.094Z"
      }

      authorized_networks {
        value           = "101.0.0.0/16"
        name            = "another"
        expiration_time = "2017-11-15T16:19:00.094Z"
      }
    }

    backup_configuration {
      enabled    = "true"
      start_time = "19:19"
    }

    activation_policy = "ON_DEMAND"
  }
}
```

I get consistent plan and apply behavior on that w/ both Terraform v0.6.12 and Terraform v0.7.0.

Those of you seeing this issue, can you see what I might be missing in trying to reproduce the bug?

**Comment by coreone** [comment]
@phinze I just tested this out with 0.7.0.  I had one authorized network that I put in on my CloudSQL instance manually.  I then added that same network as well as another VM's IP entry and ran plan/apply, and when I go back to the CloudSQL Authorization page, there are now no Authorized networks listed.  Even though the plan/apply looked to work, it actually just blanks out the networks.  This has been tested on many projects, so I know it's not a problem with this one project.  I have tried with and without expiration_time set with the same result.

When you run this with 0.7.0, do you actually see the authorized networks being populated correctly in the web interface after the apply has run?  When you run another plan after running the apply, does it say that it still needs to do the same apply again?

**Comment by phinze** [comment]
Thanks for the follow up @coreone.

> When you run this with 0.7.0, do you actually see the authorized networks being populated correctly in the web interface after the apply has run? 

Yep! With the above config: 

<img width="803" alt="tf-lw-0____access_control_-_terraform_testing" src="https://cloud.githubusercontent.com/assets/37534/17742337/3f6310aa-6465-11e6-921c-89d80fbb2d64.png">

> When you run another plan after running the apply, does it say that it still needs to do the same apply again?

With the above config, here is a follow-on empty `terraform plan`:

```
google_sql_database_instance.instance: Refreshing state... (ID: tf-lw-0612-01)

No changes. Infrastructure is up-to-date. This means that Terraform
could not detect any differences between your configuration and
the real physical resources that exist. As a result, Terraform
doesn't need to do anything.
```

> [emphasis mine] I had one authorized network **that I put in on my CloudSQL instance manually**. I then added that same network as well as another VM's IP entry and ran plan/apply, and when I go back to the CloudSQL Authorization page, there are now no Authorized networks listed. Even though the plan/apply looked to work, it actually just blanks out the networks.

Aha! This might be the key! I did not do any manual steps in the web console as I was testing. It sounds like your repro steps are:

 1. Create `google_sql_database_instance` with no authorized networks
 2. Add authorized network manually in web console
 3. Attempt to duplicate config back into Terraform
 4. Instead of resolving, Terraform clears out authorized networks

Does that sound correct?

**Comment by coreone** [comment]
@phinze Your example does indeed work, so I did some testing.  I seem to be able to break your example by doing the following:

* editing the `.tf` file to remove all the authorized_network settings
* adding an authorized network manually via the web interface
* enabling one authorized_network in the `.tf` file
* running a plan/apply, which deletes all authorized networks

From this point on, even if I take out all authorized_network entries, run a plan/apply again (to make sure Terraform's state jives with the CloudSQL state of no authorized networks), and then add authorized_networks back in, I get the continual "things need to be added" plan and the "I added those things" apply results that cycles continuously.  So, it seems that somehow touching the web interface somehow makes adding authorized_networks entries in the future impossible.

Additionally, I was also able to break this just using Terraform by:

* creating a new instance by changing the name so that Terraform would tear down the old instance and create a new one (which accurately had both authorized_networks entries after creation)
* editing the `.tf` file to remove both authorized_networks entries
* run a plan/apply and checked via the web interface that the authorized networks were in fact deleted
* add the authorized_networks entries back into the `.tf` file
* run a plan/apply, but this time the web interface (and __gcloud sql --project project-name instances describe tf-lw-0615__) both show that no authorized networks exist.

This is definitely a bug, I just haven't been able to figure out if it's on the Google API side or some internal Terraform issue, but hopefully you now have some more information to reproduce the issue.

As a side note, your test uses a first generation instance, but I have also seen the same issue with second generation instances.

**Comment by phinze** [comment]
Great, thanks for this @coreone - let me take this new data and run with it and I'll get back to you with more information.

**Comment by lwander** [comment]
I submitted a fix ^ 

I verified by

1. creating a sql instance with that template above
2. creating a new network using the cloud console
3. deleting a the networks from the template
4. running tf apply
5. verifying that the networks from the cloud console still exist
6. creating new networks in both the template and cloud console
7. running tf apply
8. verifying that all network are created

**Comment by stack72** [comment]
closed via #8290 :)
